include/rocksdb/compressor.h

```diff
@@ -0,0 +1,206 @@
+  CompressionDict& operator=(const CompressionDict&) = delete;
+  CompressionDict(CompressionDict&&) = delete;
+  CompressionDict& operator=(CompressionDict&&) = delete;
+};
+
+// Holds dictionary and related data, like ZSTD's digested uncompression
+// dictionary.
+struct UncompressionDict {
+  // Block containing the data for the compression dictionary in case the
+  // constructor that takes a string parameter is used.
+  std::string dict_;
+
+  // Block containing the data for the compression dictionary in case the
+  // constructor that takes a Slice parameter is used and the passed in
+  // CacheAllocationPtr is not nullptr.
+  CacheAllocationPtr allocation_;
+
+  // Slice pointing to the compression dictionary data. Can point to
+  // dict_, allocation_, or some other memory location, depending on how
+  // the object was constructed.
+  Slice slice_;
+
+  // If the compressor supports processing the dictionary
+  void* processed_dict_ = nullptr;
+
+  std::shared_ptr<Compressor> compressor_ = nullptr;
+
+  UncompressionDict(std::string dict, CompressionType type);
+  UncompressionDict(Slice slice, CacheAllocationPtr&& allocation,
+                    CompressionType type);
+  UncompressionDict(UncompressionDict&& rhs);
+  ~UncompressionDict();
+  UncompressionDict& operator=(UncompressionDict&& rhs);
+
+  // The object is self-contained if the string constructor is used, or the
+  // Slice constructor is invoked with a non-null allocation. Otherwise, it
+  // is the caller's responsibility to ensure that the underlying storage
+  // outlives this object.
+  bool own_bytes() const { return !dict_.empty() || allocation_; }
+  const Slice& GetRawDict() const { return slice_; }
+  const void* GetProcessedDict() const { return processed_dict_; }
+  static const UncompressionDict& GetEmptyDict() {
+    static UncompressionDict empty_dict{};
+    return empty_dict;
+  }
+  size_t ApproximateMemoryUsage() const;
+
+  UncompressionDict() = default;
+  // Disable copy
+  UncompressionDict(const CompressionDict&) = delete;
+  UncompressionDict& operator=(const CompressionDict&) = delete;
+};
+
+class CompressionContext {
+ private:
+  void* context_ = nullptr;
+  std::shared_ptr<Compressor> compressor_ = nullptr;
+
+ public:
+  explicit CompressionContext(CompressionType type);
+  ~CompressionContext();
+  CompressionContext(const CompressionContext&) = delete;
+  CompressionContext& operator=(const CompressionContext&) = delete;
+  void* GetContext() const { return context_; };
+};
+
+class CompressionInfo {
+  const CompressionOptions& opts_;
+  const CompressionContext& context_;
+  const CompressionDict& dict_;
+  const CompressionType type_;
+  const uint64_t sample_for_compression_;
+
+ public:
+  CompressionInfo(const CompressionOptions& _opts,
+                  const CompressionContext& _context,
+                  const CompressionDict& _dict, CompressionType _type,
+                  uint64_t _sample_for_compression)
```

Comment:
Rather than taking in a "CompressionOptions" and "CompressionType" should it take in a Compressor?  Why wouldn't we create. Compressor based on these arguments and return it?

Same goes for UncompressionInfo.

Reply:
CompressionOptions and CompressionType carry sufficient information and I didn't change them.

The CompressionType in CompressionInfo/UncompressionInfo is used to get a Compressor from the registry in the CompressData/UncompressData functions (where Compress/Uncompress are eventually called). It is also used in a few other functions, such as CompressBlock and UncompressBlockContents (only the type is needed there). 
A related question is whether we should get the Compressor earlier in the call chain rather than waiting till the last moment. I didn't see a benefit, and it would have triggered more code changes.

The CompressionOptions in CompressionInfo are used to pass options to the Compress method. This part didn't change. Those options are not meant to be stored in the Compressor state. As a Compressor instance is shared across all threads, applying the options to the Compressor would immediately impact any threads using it (e.g., any threads in the middle of a compaction). On the other hand, CompressionOptions are captured at the beginning of each compaction (by GetCompressionOptions) and stay unchanged for the duration of the job. 
However, there may be options that should be applied to the Compressor itself (for example, any options with effect beyond the Compress method). How to manage those options in a safe manner is something I still need to close on. It could be part of PR #6717.